snmp_exporter/collector.py

```diff
@@ -5,21 +5,42 @@
 
 from prometheus_client import Metric, CollectorRegistry, generate_latest, Gauge
 
-def walk_oids(host, port, oids, community):
-  session = netsnmp.Session(Version=2, DestHost=host, RemotePort=port,
-      Community=community, UseNumeric=True, Retries=3)
+def walk_oids(host, port, oids, version, auth):
+  if version == 3:
+    auth_user = auth['username']
+    auth_pass = auth['password']
+    auth_protocol = auth.get('auth_protocol', 'MD5')
+    priv_protocol = auth.get('priv_protocol', 'DES')
+    priv_pass = auth['priv_password']
+    sec_level = auth.get('security_level', 'noAuthNoPriv')
+    session = netsnmp.Session(Version=3, DestHost=host, RemotePort=port,
+        SecName=auth_user, AuthPass=auth_pass, AuthProto=auth_protocol,
+        PrivProto=priv_protocol, PrivPass=priv_pass, SecLevel=sec_level,
+        UseNumeric=True, Retries=3)
+  elif version == 2:
+    community = auth.get('community', 'public')
+    session = netsnmp.Session(Version=2, DestHost=host, RemotePort=port,
+        Community=community, UseNumeric=True, Retries=3)
+  elif version == 1:
+    community = auth.get('community', 'public')
+    session = netsnmp.Session(Version=1, DestHost=host, RemotePort=port,
+        Community=community, UseNumeric=True, Retries=3)
   for oid in oids:
-    for v in walk_oid(session, oid):
+    for v in walk_oid(session, version, oid):
       yield v
 
-def walk_oid(session, oid):
+def walk_oid(session, version, oid):
     last_oid = oid
     while True:
       # getbulk starts from the last oid we saw.
       vl = netsnmp.VarList(netsnmp.Varbind('.' + last_oid))
-      if not session.getbulk(0, 25, vl):
-        return
-
+      if version == 1:
+        if not session.getnext(vl):
```

Comment:
My proposal would be: https://github.com/thomseddon/snmp_exporter/commit/e6b8fe9b44613eb5053198afa7c4e0bc68885ee8 (also removes trailing white spaces)